src/coreComponents/physicsSolvers/multiphysics/PoromechanicsKernelBase.hpp

```diff
@@ -0,0 +1,266 @@
+    /// Delta temperature since last time step
+    real64 deltaTemperatureFromLastStep{}; // for porosity update
+
+    // Storage for residual and degrees of freedom
+
+    /// Linear momentum balance residual
+    real64 ( &localResidualMomentum )[numDispDofPerElem];
+    /// Derivative of linear momentum balance residual wrt displacement
+    real64 ( &dLocalResidualMomentum_dDisplacement )[numDispDofPerElem][numDispDofPerElem];
+    /// Derivative of linear momentum balance residual wrt pressure
+    real64 dLocalResidualMomentum_dPressure[numDispDofPerElem][1]{};
+
+    /// C-array storage for the element local row degrees of freedom.
+    globalIndex localPressureDofIndex[1]{};
+
+  };
+  //*****************************************************************************
+
+  /**
+   * @brief Copy global values from primary field to a local stack array.
+   * @copydoc ::geosx::finiteElement::ImplicitKernelBase::setup
+   *
+   * For the PoromechanicsKernelBase implementation, global values from the displacement,
+   * incremental displacement, and degree of freedom numbers are placed into
+   * element local stack storage.
+   */
+  GEOSX_HOST_DEVICE
+  GEOSX_FORCE_INLINE
+  void setup( localIndex const k,
+              StackVariables & stack ) const
+  {
+    for( localIndex a = 0; a < numNodesPerElem; ++a )
+    {
+      localIndex const localNodeIndex = m_elemsToNodes( k, a );
+
+      for( integer i = 0; i < 3; ++i )
+      {
+#if defined(CALC_FEM_SHAPE_IN_KERNEL)
+        stack.xLocal[a][i] = m_X[localNodeIndex][i];
+#endif
+        stack.u_local[a][i] = m_disp[localNodeIndex][i];
+        stack.uhat_local[a][i] = m_uhat[localNodeIndex][i];
+        stack.localRowDofIndex[a*3+i] = m_dofNumber[localNodeIndex]+i;
+        stack.localColDofIndex[a*3+i] = m_dofNumber[localNodeIndex]+i;
+      }
+    }
+
+    stack.localPressureDofIndex[0] = m_flowDofNumber[k];
+  }
+
+protected:
+
+  /// The array containing the nodal position array.
+  arrayView2d< real64 const, nodes::REFERENCE_POSITION_USD > const m_X;
+
+  /// The rank-global displacement array.
+  arrayView2d< real64 const, nodes::TOTAL_DISPLACEMENT_USD > const m_disp;
+
+  /// The rank-global incremental displacement array.
+  arrayView2d< real64 const, nodes::INCR_DISPLACEMENT_USD > const m_uhat;
+
+  /// The gravity vector.
+  real64 const m_gravityVector[3]{};
+  real64 const m_gravityAcceleration;
+
+  /// The rank global density
+  arrayView2d< real64 const > m_solidDensity;
+
+  /// The global degree of freedom number
+  arrayView1d< globalIndex const > const m_flowDofNumber;
+
+  /// The rank-global fluid pressure at the previous converged time step
+  arrayView1d< real64 const > const m_pressure_n;
+  /// The rank-global fluid pressure
+  arrayView1d< real64 const > const m_pressure;
+
+};
+
+} // namespace poromechanicsKernels
+
+} // namespace geosx
+
+#include "finiteElement/kernelInterface/SparsityKernelBase.hpp"
```

Comment:
This reference is always included in the bottom, I did a try to move it up and everything still work. What is the special reason of putting this include in the bottom? (This comment is not particulaly related to this PR)

Reply:
Yes, it is not necessary to include this everywhere, this was just a copy-paste, I removed it